### PR TITLE
Nominator: Allow multiple nominations per round

### PR DESCRIPTION
**Pull request by omerfirmak**
SCP rules allows nodes to nominate multiple values.
Only constraint is that once a node confirms a nomination, it should
stop nominating new values.

### CODE REVIEW

**Comment by AndrejMitrovic** [comment]
>  it should stop nominating new values

Technically the node can choose to do whatever it wants, but other (non-faulty/non-byzantine) nodes will ignore the new values if they know that node has already confirmed something.

Anyway just trying to clear it up for everyone.

**Comment by ferencdg** [comment]
I wonder what is the background of this change. At the end of the 10 mins interval when the nomination begins, the set of transactions will be fairly stable. Once we called the 'nominate' method of the SCP, then there is an internal timer inside SCP that kicks in, and tries to renominate the same set of transactions(and other node's nominations too) that were passed into the 'nominate' method for that round.

File: source/agora/consensus/protocol/Nominator.d [hunk 1]
```diff
@@ -817,6 +814,13 @@ extern(D):
                 "Error: {}", ex.msg);
             return ValidationLevel.kInvalidValue;
         }
+
+        // Per SCP rules, once we CONFIRM a NOMINATE; we can't
+        // nominate new values. Keep track of the biggest slot_idx we confirmed
+        // a NOMINATE on
+        if (slot_idx > this.done_nominating)
```
Comment (AndrejMitrovic) [inline source/agora/consensus/protocol/Nominator.d]: I don't think this is the right place to do this. It assumes the node will confirm something as soon as its validated and this function returns.

Isn't there a confirm event handler virtual method somewhere in SCP? Or something similar to that.

Comment (AndrejMitrovic) [reply source/agora/consensus/protocol/Nominator.d]: Maybe something from here?

https://github.com/stellar/stellar-core/blob/54b03f755ae5d5aa12a799c8f1ee4d87fc9d1a1d/src/scp/SCPDriver.h#L205-L240

Comment (omerfirmak) [reply source/agora/consensus/protocol/Nominator.d]: Other than `updatedCandidateValue` they look unrelated. It's called a bit further into the process, seems like a better place to do this. 

I moved it there.

**Comment by omerfirmak** [comment]
> I wonder what is the background of this change. At the end of the 10 mins interval when the nomination begins, the set of transactions will be fairly stable. Once we called the 'nominate' method of the SCP, then there is an internal timer inside SCP that kicks in, and tries to renominate the same set of transactions(and other node's nominations too) that were passed into the 'nominate' method for that round.

Flexibility mostly, we should not tie our own hands. IMO, we should take advantage of the SCP as much as we can.
If I am not mistaken, nomination phase can take arbitrarily long. So letting nodes nominate multiple values could help.

File: source/agora/consensus/protocol/Nominator.d (hunk 1 above)
Comment (omerfirmak) [reply source/agora/consensus/protocol/Nominator.d]: `emitEnvelope` would probably be even better. We could check for an outgoing CONFIRM envelope. What do you think?

Comment (omerfirmak) [reply source/agora/consensus/protocol/Nominator.d]: We can squash the last commit once we are happy with it.

**Comment by linked0** [comment]
Please make a note of the original issue from which this PR comes, like "Part of #1228" in the description of this PR. It can make us find the original issue.

**Comment by omerfirmak** [comment]
> Please make a note of the original issue from which this PR comes, like "Part of #1228" in the description of this PR. It can make us find the original issue.

Well, this is not necessarily part of any existing issue tho. It is more like an overall improvement.

File: source/agora/consensus/protocol/Nominator.d [hunk 2]
```diff
@@ -183,7 +183,6 @@ extern(D):
     public void setQuorumConfig (const ref QuorumConfig quorum,
         const(QuorumConfig)[] other_quorums) nothrow @safe
     {
-        assert(!this.is_nominating);
```
Comment (AndrejMitrovic) [inline source/agora/consensus/protocol/Nominator.d]: This was a useful assert to catch bugs.

Even though SCP by design allows quorums to change mid-nomination, in our design the quorums should be fixed within every new block height nomination.

File: source/agora/consensus/protocol/Nominator.d [hunk 3]
```diff
@@ -106,8 +106,8 @@ public extern (C++) class Nominator : SCPDriver
     /// Currently active timers grouped by type
     private ITimer[TimerType.max + 1] active_timers;
 
-    /// Whether we're in the asynchronous stage of nominating
-    private bool is_nominating;
+    /// Last height that we finished the nomination round
+    private Height done_nominating;
```
Comment (AndrejMitrovic) [inline source/agora/consensus/protocol/Nominator.d]: `done` usually implies a boolean. How about `last_confirmed_height`?

File: source/agora/consensus/protocol/Nominator.d (hunk 2 above)
Comment (omerfirmak) [reply source/agora/consensus/protocol/Nominator.d]: I have added `is_nominating` back for this.

**Comment by AndrejMitrovic** [comment]
Looks good now, just please squash.

**Comment by omerfirmak** [comment]
> Looks good now, just please squash.

Done

**Review by AndrejMitrovic — APPROVED** [review]

**Pull request merged**